Introducing the notion of shards as an encapsulation of primary and s…

gsriram7 opened the pull request:
…econdary partitions

gsriram7 committed:
Introducing the notion of shards as an encapsulation of primary and secondary partitions

gsriram7 committed:
Fixing a compilation error in CHRWithShards

gsriram7 commented:
@ashwanthkumar

gsriram7 committed:
Removing unused variables

Added a FIXME wrt follower allocation

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/partitioner/CHRWithShards.scala:
```diff
@@ -0,0 +1,52 @@
+
+import in.ashwanthkumar.suuchi.cluster.MemberAddress
+
+case class Shard(primaryPartition: VNode, follower: List[VNode]) {
```
Can we please add a notion of `id` / `name` to Shard?

ashwanthkumar replied:
For cases where we know they're primary and followers - it would be best if we can abstract those away as `MemberAddress` isn't? `VNode` doesn't make sense anymore to me. What do you think?

ashwanthkumar replied:
Also I thought we were using `partitions` and not `partitionsPerNode` ?

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/partitioner/CHRWithShards.scala:
```diff
@@ -0,0 +1,52 @@
+}
+
+class CHRWithShards(hashFn: Hash, partitionsPerNode: Int, replicationFactor: Int = 2) {
+  val sortedMap = new util.TreeMap[Integer, Shard]()
```
Can we please mark this `private`? This is exposed as `public`.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/partitioner/CHRWithShards.scala:
```diff
@@ -0,0 +1,52 @@
+  def getNodes = primaryPartition.node :: follower.map(_.node)
+}
+
+class CHRWithShards(hashFn: Hash, partitionsPerNode: Int, replicationFactor: Int = 2) {
```
It would be nice if we don't have defaults as part of the class and move them as part of the companion object or place them where it's actually being used.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/partitioner/CHRWithShards.scala:
```diff
@@ -0,0 +1,52 @@
+    this
+  }
+
+  private def hash(shard: Shard): Int = hashFn.hash(shard.key.getBytes)
```
Please move all `private` methods to the bottom of the class.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/partitioner/CHRWithShards.scala:
```diff
@@ -0,0 +1,52 @@
+import in.ashwanthkumar.suuchi.cluster.MemberAddress
+
+case class Shard(primaryPartition: VNode, follower: List[VNode]) {
+  def key = primaryPartition.node.host + "_" + primaryPartition.node.port + "_" + primaryPartition.nodeReplicaId
```
What's a `nodeReplicaId` ? Did it come in as part of #58 ?

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/partitioner/CHRWithShards.scala:
```diff
@@ -0,0 +1,52 @@
+
+  def add(node: MemberAddress, nodes: List[MemberAddress]) = {
+    // FIXME: followers based on nodes.filterNot will lead to a skew in first elements in the list
+    (1 to partitionsPerNode).map(i => Shard(VNode(node, 1, Primary), followers(nodes.filterNot(_.equals(node))))).foreach { shard =>
```
Selva, Can we please have a test for this? It doesn't look right to me.

ashwanthkumar commented:
@gsriram7 Can we also please add tests to this PR? Also can you print the CHRing with some sample data and see if you can plot it on a ring like structure by taking their ranges into account.

gsriram7 committed:
Using a priority queue approach to assign followers to shards

Added tests for NodePriority

P.S More tests need to be added to CHRWithShards and method and class names of NodePriority has to be refactored

ashwanthkumar commented:
Closing this in favor of #60
